malloc_free_check/malloc_free_check_extension.py

```diff
@@ -0,0 +1,135 @@
+"""
+Undo Automation extension module for tracking calls to malloc() and free() and checking for
+leaked memory.
+
+This script only support the x86-64 architecture.
+
+Contributors: Chris Croft-White, Magne Hov
+"""
+
+import collections
+import re
+
+import gdb
+
+from undodb.debugger_extensions import udb
+from undodb.debugger_extensions.debugger_io import redirect_to_launcher_output
+
+
+def leak_check() -> int:
+    """
+    Implements breakpoints and stops on all calls to malloc() and free(), capturing the
+    timestamp, size and returned pointer for malloc(), then confirms the address pointer is later
+    seen in a free() call.
+
+    If a subsequent free() is not seen, then at the end of execution, output the timestamp and
+    details of the memory which was never freed.
+
+    Returns the number of unmatched allocations found.
+    """
+    # Set a breakpoint for the specified function.
+    gdb.Breakpoint("malloc")
+    gdb.Breakpoint("free")
+
+    # Declare allocations dictionary structure.
+    allocations = collections.OrderedDict()
+
+    # Do "continue" until we have gone through the whole recording, potentially
+    # hitting the breakpoints several times.
+    end_of_time = udb.get_event_log_extent().end
+    while True:
```

Comment:
This is likely to break if there are signals in the recording. I guess it doesn't particularly matter, but maybe we should expose `signals_suspended` from our internal API.